internal/ingress/controller/template/template.go

```diff
@@ -882,41 +884,87 @@ func proxySetHeader(loc interface{}) string {
+func buildCustomErrorDeps(upstreamName string, errorCodes []int, enableMetrics bool) interface{} {
 	return struct {
-		ProxySetHeaders map[string]string
-		ErrorCodes      []int
-		EnableMetrics   bool
+		UpstreamName  string
+		ErrorCodes    []int
+		EnableMetrics bool
 	}{
-		ProxySetHeaders: proxySetHeaders,
-		ErrorCodes:      errorCodes,
-		EnableMetrics:   enableMetrics,
+		UpstreamName:  upstreamName,
+		ErrorCodes:    errorCodes,
+		EnableMetrics: enableMetrics,
 	}
 }
 
-// collectCustomErrorsPerServer is a utility function which will collect all
+// getDefaultBackendUpstreamNameForLocation returns the upstream-formatted name of the
+// default backend for a given location (either custom, specified by annotation, or
+// the global default)
+func getDefaultBackendUpstreamNameForLocation(input interface{}) string {
+	location, ok := input.(*ingress.Location)
+	if !ok {
+		klog.Errorf("expected a '*ingress.Location' type but %T was returned", input)
+		return ""
+	}
+
+	if location.DefaultBackend != nil {
+		return fmt.Sprintf("custom-default-backend-%v", location.DefaultBackend.GetName())
+	}
+	return "upstream-default-backend"
+}
+
+type errorLocation struct {
+	Name  string
+	Codes []int
+}
+
+// buildCustomErrorLocationsPerServer is a utility function which will collect all
 // custom error codes for all locations of a server block, deduplicates them,
-// and returns a unique set (for the template to create @custom_xxx locations)
-func collectCustomErrorsPerServer(input interface{}) []int {
+// and returns a set which is unique by default-upstream and error code. The resulting
+// data structure is a map of upstreamName => []errorCodes, so that sufficiently unique
+// @custom error location blocks can be created in the template
+func buildCustomErrorLocationsPerServer(input interface{}) interface{} {
 	server, ok := input.(*ingress.Server)
 	if !ok {
 		klog.Errorf("expected a '*ingress.Server' type but %T was returned", input)
 		return nil
 	}
 
-	codesMap := make(map[int]bool)
+	codesMap := make(map[string]map[int]bool)
 	for _, loc := range server.Locations {
+		backendUpstream := getDefaultBackendUpstreamNameForLocation(loc)
+
+		var dedupedCodes map[int]bool
+		if existingMap, ok := codesMap[backendUpstream]; ok {
+			dedupedCodes = existingMap
+		} else {
+			dedupedCodes = make(map[int]bool)
+		}
+
 		for _, code := range loc.CustomHTTPErrors {
-			codesMap[code] = true
+			dedupedCodes[code] = true
 		}
+		codesMap[backendUpstream] = dedupedCodes
 	}
 
-	uniqueCodes := make([]int, 0, len(codesMap))
-	for key := range codesMap {
-		uniqueCodes = append(uniqueCodes, key)
+	errorLocations := []errorLocation{}
+
+	for upstream := range codesMap {
+		codesForUpstream := []int{}
+		for code := range codesMap[upstream] {
+			codesForUpstream = append(codesForUpstream, code)
+		}
+		sort.Ints(codesForUpstream)
+		errorLocations = append(errorLocations, errorLocation{
+			Name:  upstream,
+			Codes: codesForUpstream,
```

Comment:
Shoulf we sort the codes as well? Since you're using range at https://github.com/kubernetes/ingress-nginx/pull/3730/files#diff-652deabba49a2a87d4c5a79700dfa1e3R953 when collecting codes, I think their order can change and introduce unnecessary Nginx reloads.

Reply:
Unless I made a mistake, that's what I was doing with `sort.Ints()` here: https://github.com/kubernetes/ingress-nginx/pull/3730/files#diff-652deabba49a2a87d4c5a79700dfa1e3R956

Reply:
My bad, yes you're already doing it 👍